scipy/optimize/tnc/_tnc_module.pyx

```diff
@@ -0,0 +1,205 @@
+# cython: language_level=3, boundscheck=False
+
+from libc.string cimport memcpy
+import numpy as np
+cimport numpy as np
+cimport cython
+
+
+np.import_array()
+
+ctypedef np.float64_t float64_t
+
+
+cdef extern from "tnc.h":
+    ctypedef void tnc_callback(double[], void*) except *
+    ctypedef int tnc_function(double[], double*, double[], void*) except 1
+
+    cdef enum tnc_rc:
+        TNC_MINRC
+        TNC_ENOMEM
+        TNC_EINVAL
+        TNC_INFEASIBLE
+        TNC_LOCALMINIMUM
+        TNC_FCONVERGED
+        TNC_XCONVERGED
+        TNC_MAXFUN
+        TNC_LSFAIL
+        TNC_CONSTANT
+        TNC_NOPROGRESS
+        TNC_USERABORT
+
+    cdef int tnc(int n, double x[], double *f, double g[],
+            tnc_function *function, void *state,
+            double low[], double up[], double scale[], double offset[],
+            int messages, int maxCGit, int maxnfeval, double eta, double stepmx,
+            double accuracy, double fmin, double ftol, double xtol, double pgtol,
+            double rescale, int *nfeval, int *niter, tnc_callback *tnc_callback) except *
+
+
+cdef struct s_pytnc_state:
+  void *py_function
+  void *py_callback
+  int n
+  int failed
+ctypedef s_pytnc_state pytnc_state
+
+
+cdef int function(double x[], double *f, double g[], void *state) except 1:
+    # return 0 if no error, 1 if error
+    # However, if the user function raises Exceptions Cython should deal
+    # with it.
+
+    cdef:
+        pytnc_state *py_state
+        int n
+        double *x_data
+        double *g_data
+
+    py_state = <pytnc_state *>state
+    n = py_state.n
+
+    # ensures we're working on a copy of the data, just in case user function
+    # mutates it
+    xcopy = np.empty(n, dtype=float)
+    x_data = <float64_t *>np.PyArray_DATA(xcopy)
+    memcpy(x_data, x, sizeof(double) * n)
+
+    fx, gx = (<object>py_state.py_function)(xcopy)
+
+    if not np.isscalar(fx):
+        try:
+            fx = np.asarray(fx).item()
+        except (TypeError, ValueError) as e:
+            raise ValueError(
+                "The user-provided objective function "
+                "must return a scalar value."
+            ) from e
+
+    f[0] = <double> fx
+
+    gx = np.asfarray(gx)
+    if gx.size != n:
+        raise ValueError("tnc: gradient must have shape (len(x0),)")
+
+    if not gx.flags['C_CONTIGUOUS']:
+        gx = np.ascontiguousarray(gx, dtype=np.float64)
+
+    g_data = <float64_t *>np.PyArray_DATA(gx)
+    memcpy(g, g_data, n * sizeof(double))
+
+    return 0
+
+
+cdef void callback_function(double x[], void *state) except *:
+    cdef:
+        pytnc_state *py_state
+        int n
+        double *x_data
+
+    py_state = <pytnc_state *>state
+    n = py_state.n
+
+    # ensures we're working on a copy of the data, just in case user function
+    # mutates it
+    try:
+        xcopy = np.empty(n, dtype=float)
+        x_data = <float64_t *>np.PyArray_DATA(xcopy)
+        memcpy(x_data, x, sizeof(double) * n)
```

Comment:
This may not be necessary - it looks from tnc.c like `callback_function` receives a temporary copy of the data anyway.

But I could be misunderstanding.

Reply:
`tnc.c` has to use a temporary array because the in-place unscaling/scaling is altering `x` by a small amount. So it's necessary to memcpy to the temp array, unscale in-place, then callback with that temp array.

The comment here may be a little misleading. Here we need to call a Python function with the contents of `x`. Is there a more direct way of doing that than creating a numpy array, transferring the information in `x` to that array, then calling the Python callback function?

Reply:
If you don't need to make a copy, then potentially you could use Cython memoryviews:

```
xview = np.asarray(<double[:n]>(x))  # probably don't use this code - see below!
```

That creates a Cython memoryview of `x` and then creates a Numpy array that wraps the view.

**However**, neither the Numpy array or the Cython memoryview actually own the memory. Therefore, if a user decides to store that array inside the callback (to plot it later, for example) then it will be invalid by the time they want it. 

I think your "create a Numpy array and copy `x` into it is probably the right thing to do.

Reply:
Just to summarise, the upshot of all of that is that I think you should keep making the copy (like you are doing now)